File: impl/ts/src/tool/glob.ts
// glob — find project files by pattern. See docs/harness-design.md §5.2.
// Walks the jailed tree (walk.ts) and matches a glob pattern (**, *, ?) without
// pulling in a dependency. Bounded to 100 results like opencode.

import type { ToolDef, ToolContext, ToolResult } from "./types.js";
import { reqString, optString } from "./validate.js";
import { walkFiles } from "./walk.js";

const LIMIT = 100;

interface Args {
  pattern: string;
  path?: string;
}

// Translate a glob to a RegExp. Order matters: handle ** before *.
function globToRe(glob: string): RegExp {
  let re = "";
  for (let i = 0; i < glob.length; i++) {
    const c = glob[i];
    if (c === "*") {
      if (glob[i + 1] === "*") {
        // ** matches across path separators (and an optional trailing /).
        re += ".*";
        i++;
        if (glob[i + 1] === "/") i++;
      } else {
        re += "[^/]*";
      }
    } else if (c === "?") {
      re += "[^/]";
    } else if (".+^${}()|[]\\".includes(c)) {
      re += "\\" + c;
    } else {
      re += c;
    }
  }
  return new RegExp("^" + re + "$");
}

export const GlobTool: ToolDef<Args> = {
  id: "glob",
  description:
    "Find files by glob pattern (e.g. src/**/*.tsx, **/Root.tsx). Returns matching project-relative paths. Use to locate files before reading them.",
  parameters: {
    pattern: { type: "string", description: "Glob pattern (** matches any depth)", required: true },
    path: { type: "string", description: "Directory to search under (default project root)" },
  },
  validate(args) {
    return { pattern: reqString(args, "pattern"), path: optString(args, "path") };
  },
  async execute(args, ctx: ToolContext): Promise<ToolResult> {
    const root = args.path ?? ".";
    const re = globToRe(args.pattern);
    const all = walkFiles(ctx, root);
    const matched = all.filter((p) => re.test(p));
    const truncated = matched.length > LIMIT;
    const shown = matched.slice(0, LIMIT);
    let output = shown.length === 0 ? "No files found" : shown.join("\n");
    if (truncated) {
      output += `\n\n(Results truncated: showing first ${LIMIT} of ${matched.length}. Use a more specific pattern.)`;
    }
    return { title: args.pattern, output, metadata: { count: shown.length, truncated } };
  },
};

File: impl/go/internal/tool/glob.go
// glob — find project files by pattern (Go parity of glob.ts).
package tool

import (
	"fmt"
	"regexp"
	"strings"
)

const globLimit = 100

// globToRe translates a glob (**, *, ?) to a regexp anchored to the whole path.
func globToRe(glob string) *regexp.Regexp {
	var b strings.Builder
	b.WriteString("^")
	for i := 0; i < len(glob); i++ {
		c := glob[i]
		switch {
		case c == '*':
			if i+1 < len(glob) && glob[i+1] == '*' {
				b.WriteString(".*")
				i++
				if i+1 < len(glob) && glob[i+1] == '/' {
					i++
				}
			} else {
				b.WriteString("[^/]*")
			}
		case c == '?':
			b.WriteString("[^/]")
		case strings.ContainsRune(".+^${}()|[]\\", rune(c)):
			b.WriteByte('\\')
			b.WriteByte(c)
		default:
			b.WriteByte(c)
		}
	}
	b.WriteString("$")
	re, err := regexp.Compile(b.String())
	if err != nil {
		return regexp.MustCompile("^$")
	}
	return re
}

// GlobTool finds files by glob pattern.
var GlobTool = Def{
	ID:          "glob",
	Description: "Find files by glob pattern (e.g. src/**/*.tsx, **/Root.tsx). Returns matching project-relative paths. Use to locate files before reading them.",
	Parameters: []Param{
		{Name: "pattern", Type: "string", Description: "Glob pattern (** matches any depth)", Required: true},
		{Name: "path", Type: "string", Description: "Directory to search under (default project root)"},
	},
	Validate: func(args map[string]any) (map[string]any, error) {
		pat, err := reqString(args, "pattern")
		if err != nil {
			return nil, err
		}
		p, err := optString(args, "path")
		if err != nil {
			return nil, err
		}
		return map[string]any{"pattern": pat, "path": p}, nil
	},
	Execute: func(args map[string]any, ctx *Context) (Result, error) {
		pattern := args["pattern"].(string)
		root := args["path"].(string)
		if root == "" {
			root = "."
		}
		re := globToRe(pattern)
		var matched []string
		for _, p := range walkFiles(ctx, root, 5000) {
			if re.MatchString(p) {
				matched = append(matched, p)
			}
		}
		truncated := len(matched) > globLimit
		shown := matched
		if truncated {
			shown = matched[:globLimit]
		}
		output := "No files found"
		if len(shown) > 0 {
			output = strings.Join(shown, "\n")
		}
		if truncated {
			output += fmt.Sprintf("\n\n(Results truncated: showing first %d of %d. Use a more specific pattern.)", globLimit, len(matched))
		}
		return Result{Title: pattern, Output: output, Metadata: map[string]any{"count": len(shown), "truncated": truncated}}, nil
	},
}
